### PR TITLE
Create Asset Browser Inspector Panel

### LINKED ISSUE #12872: 17 - Create Asset Browser inspector panel
**Use Story:**

**Migrate and create an associated asset panel in the asset browser inspector panel**

As a user, I need the ability to see my asset related details in it's own panel that is not integrated tightly into the asset browser. Currently the integrated asset preview and details of that asset in Asset Browser makes the viewing experience extremely hard to see thumbnail and or details because too many things are going on in one small panel. It also make it difficult for user to pick any other different configurations for the AB beside along the bottom row. We would like to move the preview and details to a new panel. This is a current standard among all the other game engines we would like to support. See other examples of this below in the ticket.

**AC**
 * By default, the panel should be docked with entity inspector and is pulled to the front based on when an asset from asset browser is selected.
 * We should also do this in reverse that when any entity / prefab in Outliner or Viewport is selected we should pull the entity inspector to the front.
 * This new panel should have 4 main sections and 3 accordions.
 **** Section 1 - The preview image panel
 **** Section 1 - main property details (name, size, type, dimensions, last modified, date create, status, and repo)
 **** Section 2 -  secondary details - editable  ( these can be configurable based on selected asset type)
 ***** file location
 ***** Source tool it came from
 ***** Comments user would like to include for game team.
 ***** FBX related settings (later)
 ***** PhysX related settings (later)
 **** Section 3 - related assets.  
 **** Section 4 -
 ***** "Tagging" assets for searching. This will allow user to define labels for an asset for easy discovery later. Is this object a [--- blank---] type of asset. Example Vegetation, weapon, particle, animation, etc.

 

image

 

image

### LINKED ISSUE #12873: 18 - Asset Browser - Dependent assets
**Use Story1:**

As a user, I need the ability to see my associated asset to my currently selected asset in the Asset Browser inspector panel. By default, these should not be "engine processed asset (AZ--- assets), but asset as a user I would want to know about for manipulation purposes.

This includes any associated:
 * FBX
 * Matierals
 * Script Canvas files
 * Mesh (Obj)
 * Prefab
 * ProcPrefabs

"Engine related assets" should be off by default and can be toggled on if requested (see toggle). Currently these related assets can be seen as a tree arrow in the main AB. These include az"—" related assets. (azModel, azbuffer, azlod, etc).

**Dependence:** 

This ticket has the dependencies of
 * Creating the new Asset Browser inspector Panel
 * Migrating the inspector panel preview details and engine asset list into the new inspector panel.

 

AC
 * Create a new accordion panel inside the new AB inspector. This new accordion tab should say Asssoicated assets"
 * Create a new toggle switch for "related engine assets".  and hide these by default. As a Game builder, I don't want you to show me AP files unless I check a checkbox that states specifically for these details.
 * Create a tree structure and list out all known assoicated assets. This could be find via prefab level, level, metadata, etc/
 * Some option should be given when clicking on the list of the assets. These options are still TBD.

image

Image

**As a user What I want to know about is**:

IF I'm viewing a prefab/entity/material/script/ I need to the actual file being used.
 * I need to know which other prefabs are dependent on me
 * I need to know which other FBX are dependent on me.
 * I need to know which script are attached directly to me.
 * I need to know which level I'm in
 * I need to know which level

!Dependent Assets.png|thumbnail!

### PR DESCRIPTION
**Pull request by dtamkin1**
## What does this PR do?

This PR populates the new Asset Browser Inspector Panel with 3 sections:

- An asset image preview if there is one
-  An "Asset Details" card that displays more information about a selected asset, if the information is available:
    - Name
    - Location (File path)
    - File Type (fbx, image, model, etc)
    - Asset Type (source or product)
    - Disk Size (KB)
    - Dimensions
    - Vertices
- A "Dependent Assets" card that displays relationships between assets in the asset browser:
  - Products displays all of the products of a source asset
  - Outgoing Source Dependencies lists all source assets whose products are depended on by any of the products of the selected asset.
  - Incoming Source Dependencies lists all source assets whose products depend on any of the products of the selected asset.
  - Outgoing Product Dependencies lists all products that the selected product depends on
  - Incoming Product Dependencies lists all products that depend on the selected product.

Fixes #12872 
Fixes #12873 

Editor_LOq3lfn8gO

## How was this PR tested?

Locally

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserBus.h [hunk 1]
```diff
@@ -337,9 +337,12 @@ namespace AzToolsFramework
         public:
             static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Multiple;
 
-            //! Updates the asset preview panel with data about the passed entry.
+            //! Updates the asset browser inspector panel with data about the passed entry.
             //! Clears the panel if nullptr is passed
             virtual void PreviewAsset(const AzToolsFramework::AssetBrowser::AssetBrowserEntry* selectedEntry) = 0;
+
+            //! Clears the asset browser inspector panel if it has data in it
+            virtual void ClearPreview(){};
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserBus.h:345]: Can it be a pure virtual function? Also, I think the comment is not accurate here. "if it has data in it" can be removed.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h [hunk 2]
```diff
@@ -8,9 +8,16 @@
...
 
 #if !defined(Q_MOC_RUN)
-#include <QDockWidget>
+#include <AzCore/std/containers/set.h>
 #include <AzToolsFramework/AssetBrowser/AssetBrowserBus.h>
-#include <AzCore/Asset/AssetCommon.h>
+#include <AzToolsFramework/AssetDatabase/AssetDatabaseConnection.h>
+#include <AzToolsFramework/AssetBrowser/Entries/ProductAssetBrowserEntry.h>
+#include <AzToolsFramework/AssetBrowser/Entries/SourceAssetBrowserEntry.h>
+#include <AzQtComponents/Components/Widgets/Card.h>
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h:16]: nit: include order - 'AssetBrowser' should be before 'AssetDatabase'

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Views/AssetBrowserViewUtils.cpp [hunk 3]
```diff
@@ -550,20 +550,22 @@ namespace AzToolsFramework
             }
 
             // Check if this entry has a custom previewer, if so use that thumbnail
-            AZ::EBusAggregateResults<const PreviewerFactory*> factories;
-            PreviewerRequestBus::BroadcastResult(factories, &PreviewerRequests::GetPreviewerFactory, entry);
-            for (const auto factory : factories.values)
+            if (!returnIcon)
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Views/AssetBrowserViewUtils.cpp:553]: Didn't quite follow what 'returnIcon' mean here and why it is default to false. Would be good to add some comment in the interface.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h [hunk 4]
```diff
@@ -26,11 +33,32 @@ namespace AzToolsFramework
...
+            // AssetBrowserPreviewRequestBus
+            //////////////////////////////////////////////////////////////////////////
             void PreviewAsset(const AzToolsFramework::AssetBrowser::AssetBrowserEntry* selectedEntry) override;
+            void ClearPreview() override;
 
+            void PopulateSourceDependencies(const SourceAssetBrowserEntry* sourceEntry, AZStd::vector<const ProductAssetBrowserEntry*> productList);
+            bool PopulateProductDependencies(const ProductAssetBrowserEntry* productEntry);
+            void CreateSourceDependencyTree(AZStd::set<AZ::Uuid> sourceUuids, bool isOutgoing);
+            void CreateProductDependencyTree(AZStd::set<AZ::Data::AssetId> dependencyUuids, bool isOutgoing);
+            void AddAssetBrowserEntryToTree(const AssetBrowserEntry* entry, QTreeWidgetItem* headerItem);
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h:49]: A bunch of new public non-override APIs were introduced here. Would be good to add some comments.

Also, it looks like they can be private functions.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.cpp [hunk 5]
```diff
@@ -6,41 +6,461 @@
...
+            emptyLabel->setText(QObject::tr("Select an asset to view its properties in the inspector."));
+            emptyLayout->addWidget(emptyLabel);
+            emptyLayout->addStretch(1);
+            m_emptyLayoutWidget->setLayout(emptyLayout);
+
+            // Instantiate the populated previewer
+            m_populatedLayoutWidget = new QWidget(this);
+            auto populatedLayout = new QVBoxLayout(m_populatedLayoutWidget);
+
+             // Create the layout for the asset icon preview
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.cpp:40]: nit: extra space

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h [hunk 6]
```diff
@@ -26,11 +33,32 @@ namespace AzToolsFramework
...
+            bool m_dbReady = false;
+            QLabel* m_previewImage = nullptr;
+            QStackedLayout* m_layoutSwitcher = nullptr;
+            QWidget* m_emptyLayoutWidget = nullptr;
+            QWidget* m_populatedLayoutWidget = nullptr;
+            QFormLayout* m_assetDetailLayout = nullptr;
+            QWidget* m_assetDetailWidget = nullptr;
+            QTreeWidget* m_dependentProducts = nullptr;
+            AzQtComponents::Card* m_dependentAssetsCard = nullptr;
+            QFont m_headerFont;
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h:61]: Some suggestions:
1. `m_dbReady` can be renamed to `m_databaseReady`
2. I think it is more readable and easier to maintain if they can be grouped either by Qt types or by their functionalities.

**Review by forkercat — COMMENTED** [review]
Very nice addition to the Asset Browser Inspector!

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserBus.h (hunk 1 above)
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserBus.h:345]: Thanks, fixed the comment! The `QComponentEntityEditorInspectorWindow` instantiates this class, and to be a pure virtual function the `ClearPreview()` method would have to be implemented in that class. It doesn't need to be used there, so I made it a virtual function instead.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.cpp (hunk 5 above)
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.cpp:40]: Fixed!

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.cpp [hunk 7]
```diff
@@ -6,41 +6,461 @@
...
+                    }
+                    return true;
+                });
+
+            CreateProductDependencyTree(outgoingDependencyIds, true);
+            CreateProductDependencyTree(incomingDependencyIds, false);
+            return (!outgoingDependencyIds.empty() || !incomingDependencyIds.empty());
+        }
+
+        void AssetBrowserEntityInspectorWidget::CreateSourceDependencyTree(AZStd::set<AZ::Uuid> sourceUuids, bool isOutgoing)
```
[the remark this answers is not in the document]
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.cpp:393]: Thanks for the suggestion, fixed

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h (hunk 2 above)
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h:16]: Good catch, fixed

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h (hunk 4 above)
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h:49]: Made these functions private and added comments to give more context

**Review by lsemp3d — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h (hunk 6 above)
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserEntityInspectorWidget.h:61]: Changed m_dbReady to m_databaseReady, and grouped the variables by their functionalities

**Review by forkercat — APPROVED** [review]

**Pull request merged**